hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeHeartbeatDispatcher.java

```diff
@@ -99,6 +99,9 @@ public SCMDatanodeHeartbeatDispatcher(NodeManager nodeManager,
       commands = nodeManager.getCommandQueue(dnID);
 
     } else {
+      // Get the datanode details again from node manager with the topology info
+      // for registered datanodes.
+      datanodeDetails = nodeManager.getNode(datanodeDetails.getIpAddress());
```

Comment:
We should not rely on the IP address of datanode in NetworkTopology path, instead we should use the datanode UUID. It is possible that more than one datanode process is running on the same machine.

Reply:
More than one DN instances on the same machine are most likely from test/dev environment such as MiniOzoneCluster. In production, even containers in K8S has dedicate IPs.

Reply:
But IP address can change for the same datanode. In fact, we have a Jira to remove it in the future from the yaml file: HDDS-1480

Reply:
> More than one DN instances on the same machine are most likely from test/dev environment such as MiniOzoneCluster. In production, even containers in K8S has dedicate IPs.

I agree, but the problem here is that after this change the test/dev environment where there are more than one datanode process running in same machine will not even work properly. Heartbeat from different datanode process (running on same machine) will be mapped to a single process and all the other datanode process will be marked as dead even though they are heartbeating.

Reply:
@nandakumar131, yes. We will need to handle this case for the minicluster based tests.
The current topology awareness is based on a map of ip/dns->location, I think change it to uuid->location should work as long we have a mapping from uuid->ip/dns maintained.

Reply:
+1 on changing it to `uuid -> location` and maintaining a map for `uuid -> ip/dns`.